docs/spec/consensus/fork-detection.md

```diff
@@ -0,0 +1,93 @@
+of interest by downloading and verifying corresponding block headers. To be able to detect a fork, we assume an additional 
+module (fork detection module) that runs in parallel with the core lite client module, and whose role is fork detection. 
+The fork detection module maintains a list of peers (full nodes) and it periodically establishes a connection to some peer 
+(called witness) and verify correctness of headers downloaded by main module (by cross-checking with headers downloaded
+from witnesses). In case there are two different headers for the same height signed by overlapping validator sets,
+then a client downloads corresponding commit and creates a proof of valid fork.
+
+## Identifying faulty validators 
+
+Once an honest lite client detects a valid fork it needs to submit it to an honest full node for further processing so
+that faulty validators can be detected and punished. For the purpose of this document we assume that a lite client
+is able to submit a proof of fork to an honest full node in reasonable time (how we can ensure that will be addressed in a
+separate document). 
+
+For the purpose of this document we assume existence of the following information for each commit: hash of the block, 
+height, round, validator set and a set of signatures by the processes in the given validator set. An honest full node starts
+the procedure that detects misbehaving validators from the following information: for some height *h*, there are two
+conflicting commits *C1* and *C2*, where *C1* is a commit from the main chain. The detection procedure works like this:  
+
+1) if there are processes from the set *signers(C2)* that are not part of *C1.valset*, they misbehaved as they are signing 
+   protocol messages in heights they are not validators. In this case commit *C2* is a self-contained evidence of misbehavior 
+   for those processes that can be simply verified by every honest validators without additional information.  
+   For processes from the set *signers(C2)* that are part of *C1.valset* we need additional checks:
+
+2) if *C1.round == C2.round*, and some processes signed different precommit messages in both commits, then it is an 
+   equivocation misbehavior. Similarly as above, commits *C1* and *C2* are an evidence of misbehavior. Note that in this case
+   there might be additional misbehavior, but its detection require more complex detection procedure that we explain next.  
+
+3) if *C1.round != C2.round* we need to run full detection procedure. We assume (in the first version) that the full detection
+   procedure is executed by a centralized, trusted component called monitor that will be processing proof of forks and 
+   identifying faulty processes (and optionally generating evidences of misbehavior). Monitor is an honest full node that runs 
+   on the main chain. Monitor is triggered by receiving two conflicting commits where *C1.round != C2.round*. The procedure starts
```

Comment:
The latter. I was trying to avoid monitor running on the chain. This is just saying that monitor has access to blockchain data that is currently considered to be valid chain. Otherwise it's very hard to talk about faulty behaviour as there is no reference point.